torch/_dynamo/output_graph.py

```diff
@@ -1528,7 +1532,63 @@ def compile_and_call_fx_graph(self, tx, rv, root):
 
             counters["stats"]["unique_graphs"] += 1
             # This is safe because we pre-process name to be unique
-            self.install_global_unsafe(name, compiled_fn)
+            if specializations := old_fake_mode.shape_env.specializations:
+                specialization_guards = []
+                specialization_cache: dict[Specialization, Callable[[Any], Any]] = {}
+                preserved_graphargs = [
+                    replace(node.meta["grapharg"], _example=None)
+                    for node in self.placeholders
+                ]
+                sources = [a.source for a in self.graphargs]
+                for specialization in specializations:
+                    source_index = sources.index(specialization.source)
+                    check_fn_source = inspect.getsource(specialization.check_fn).strip()
+                    check_fn = guards.LAMBDA_GUARD(  # type: ignore[attr-defined]
+                        specialization.check_fn,
+                        [check_fn_source],
+                    )
+
+                    log.debug(
+                        "Compiling backend specialized graph with specialization=%s",
+                        check_fn_source,
+                    )
+
+                    specialization_guards.append(
+                        (
+                            functools.partial(
+                                lambda idx, args, check_fn=check_fn: check_fn(
+                                    args[idx]
+                                ),
+                                source_index,
+                            ),
+                            specialization,
+                        )
+                    )
+
+                @torch._dynamo.disable(reason="do not trace Dynamo-compiled graph")
+                def specialized_dispatch(*args, **kwargs):
+                    for check_fn, specialization in specialization_guards:
+                        if check_fn(args):
+                            if specialization in specialization_cache:
+                                return specialization_cache[specialization](
+                                    *args, **kwargs
+                                )
+                            for node, grapharg, arg in zip(
+                                self.placeholders, preserved_graphargs, args
+                            ):
+                                node.meta["grapharg"] = replace(grapharg, _example=arg)
+                            with self.shape_env.patch_source_specialization(
```

Comment:
cc @jamesjwu will this be a problem for aotautograd cache? ShapeEnvs aren't serializable so I imagine you using a whitelist of fields? Do you by default serialize the axioms? I'll take a look shortly but if you know the answer off the top of your head, please lmk.

Reply:
Hmmm reading the code it seems like the cache key doesn't include either guards or ShapeEnv? Instead we actually do the guard check post cache fetch? What's the reason for doing it this way over including (at least the dynamic shape guards + shapeenv axioms) in the cache key?

Reply:
Guards aren't serializable; the shape env, therefore, isn't serializable. Instead, we produce a guard expression that can be evaluated on cache hit load. 

Don't have context on your specific change here, but any change that affects the output graph needs to be reflected by either the cache key, or `shape_env.produce_guards_expression`. 

This looks like it's happening in dynamo land, though; will this change not be reflected in the output graph dynamo generates? I.e. why is AOTAutograd affected?

Reply:
```
What's the reason for doing it this way over including (at least the dynamic shape guards + shapeenv axioms) in the cache key?
```
A separate thing: AOTAutograd and Inductor can generate more guards and modify the shape env **during** compilation, so it's not sufficient to include the shape env at cache key generation time, as we may add more to the shape env afterward.

Reply:
I am also worried about James point that the aot and inductor can add extra guards. And is that possible that those guards collide with what specialization somehow (for the generic case I guess)

Reply:
While that’s true in general, in the context of a non-changing dynamo graph, I can’t think of a single case where a guard would be added post dynamo where the only difference between compiles is a narrower value range for a symbol.

Reply:
Sounds good, happy to chat about it in person next week!